### PR TITLE
dynamodb-local: init at 2023-12-14

### PR DESCRIPTION
**Pull request by shyim**
## Description of changes

Dynamodb local allows to simulate a Dynamodb completely locally and is ideal for local development.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    description = "DynamoDB Local is a small client-side database and server that mimics the DynamoDB service.";
+    homepage = "https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html";
+    license = licenses.unfree;
+    maintainers = with maintainers; [ shyim ];
+    platforms = platforms.all;
+    sourceProvenance = with lib.sourceTypes; [
+      binaryBytecode
+      binaryNativeCode
+    ];
+  };
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:38]: Missing `meta.mainProgram`

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{ lib
+, stdenv
+, fetchurl
+, jre
+, makeWrapper
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:5]: Use `makeBinaryWrapper` instead, which offers better compatibility with Darwin archs.

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share/dynamo-db-local
+    cp -r DynamoDBLocal* $out/share/dynamo-db-local
+
+    makeWrapper ${jre}/bin/java $out/bin/dynamo-db-local \
+      --add-flags "-jar $out/share/dynamo-db-local/DynamoDBLocal.jar"
+  '';
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:26]: Missing phase hooks

**Review by drupol — CHANGES_REQUESTED** [review]
Hi there,

Thanks for your submission.

Let me know if my feedback is clear.

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  nativeBuildInputs = [ makeBinaryWrapper ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin $out/share/dynamo-db-local
+    cp -r DynamoDBLocal* $out/share/dynamo-db-local
+
+    makeBinaryWrapper ${jre}/bin/java $out/bin/dynamo-db-local \
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:26]: How about keeping the same name as in `pname` ?

Basically, replacing `dynamo-db-local` by `dynamodb-local`.

Comment (shyim) [reply pkgs/by-name/dy/dynamodb-local/package.nix:26]: makes sense 👍

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  nativeBuildInputs = [ makeBinaryWrapper ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin $out/share/${finalAttrs.pname}
+    cp -r DynamoDBLocal* $out/share/${finalAttrs.pname}
+
+    makeBinaryWrapper ${jre}/bin/java $out/bin/${finalAttrs.pname} \
+      --add-flags "-jar $out/share/${finalAttrs.pname}/DynamoDBLocal.jar"
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:27]: Please don't rely on `pname`, this creates a binding that goes too far. It's better to use a string literal.

Comment (shyim) [reply pkgs/by-name/dy/dynamodb-local/package.nix:27]: also in mainProgram write the literal?

Comment (drupol) [reply pkgs/by-name/dy/dynamodb-local/package.nix:27]: Yes please

**Review by drupol — APPROVED** [review]
Thanks!

File: pkgs/by-name/dy/dynamodb-local/package.nix [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{ lib
+, stdenv
+, fetchurl
+, jre
+, makeBinaryWrapper
+}:
+stdenv.mkDerivation (finalAttrs: {
```
Comment (drupol) [inline pkgs/by-name/dy/dynamodb-local/package.nix:7]: I'm wondering if we can use `stdenNoCC` ?

Comment (shyim) [reply pkgs/by-name/dy/dynamodb-local/package.nix]: tested it works well

**Review by drupol — APPROVED** [review]
:100: Thanks :)

**Pull request merged**